### ISSUE TITLE
Testing

### ISSUE THREAD

**Issue by mhausenblas** [aws/aws-controllers-k8s#6]
This design issue is about the ACK testing setup. Usage of requirements key words as per [RFC 2119](https://tools.ietf.org/html/rfc2119).

## Scope

One of our ACK [tenets](https://github.com/aws/containers-roadmap/issues/456#issue-487568033) is that it is optimized for production usage. This requires that we have a set of test suites in place. In scope for this issue is the selection of the nature of the tests carried out as well as the design of the test infrastructure.

## Input

Suggested tests include (but are not limited to):

* Unit-level tests in Go.
* Per-AWS-service test: checks to what extent an ACK for a certain service covers the operations supported by said service.
* Performance and scale tests: to figure out the ACK behaviour under short-term load.
* Soak tests: to figure out the ACK behaviour under long-term load.

## Requirements

1. Tests in ACK `MUST` be automated. All results `SHOULD` be surfaced as bot messages on GitHub issues.
1. Every ACK feature graduating to GA `MUST` have full test coverage.
1. The selection of tests, documentation thereof, as well as the test infrastructure `MUST` be completed before implementation begins.

**Comment by daviddyball** [comment]
* Is this project sponsored? Can AWS provide testing infrastructure/credit to run CI for this?
* Do we want to rely on mocks for some of the tests? Will everything be run "for real" in AWS?
* Is there Slack infrastructure setup for this?

**Comment by jaypipes** [comment]
>     * Is this project sponsored? Can AWS provide testing infrastructure/credit to run CI for this?

Yes, e2e tests will be run on AWS infrastructure and will gate commits to the repo.

>     * Do we want to rely on mocks for some of the tests? Will everything be run "for real" in AWS?

Both :) Fast/unit tests with mocks are good. e2e functional integration tests against real AWS resources are also good.

>     * Is there Slack infrastructure setup for this?

Not yet :) Would you recommend a separate channel or use an existing one?

**Comment by iAnomaly** [comment]
> Not yet :) Would you recommend a separate channel or use an existing one?
I'd like to see a dedicated #aws-service-operator channel on the official kubernetes.slack.com Slack.

**Comment by sfzylad** [comment]
One of the testing scenarios that could be reassuring for production use would be testing how the operator reacts if a resource in AWS differs from what is stored in apiserver.

**Comment by mhausenblas** [comment]
Thanks a lot @sfzylad for this comment and can you maybe expand a little on this? What exactly do you mean with "if a resource in AWS differs from what is stored in apiserver."? As in, one manually `kubectl`s something? How would you test it? Can you come up with a concrete example that demonstrates this in action? Again, thanks and looking forward learning more.

**Comment by marcindulak** [comment]
> One of the testing scenarios that could be reassuring for production use would be testing how the operator reacts if a resource in AWS differs from what is stored in apiserver.

I'm guessing this question is about how "drift" (in cloudformation terms) of the aws resources will be handled. What will happen if I configure a resource in aws using aso, but then someone modifies it manually using aws console or by other means.

**Comment by sfzylad** [comment]
>I'm guessing this question is about how "drift" (in cloudformation terms) of the aws resources will be handled. What will happen if I configure a resource in aws using aso, but then someone modifies it manually using aws console or by other means.

That's correct.

**Comment by sfzylad** [comment]
@mhausenblas As @marcindulak mention. The resource is configured and managed by ASO and at the same time someone makes a change to this resource in the console. The test should check if the operator rolled back the resource to the declared state.

**Comment by dcherman** [comment]
@sfzylad @mhausenblas Likely out of scope for this issue, but I'd also argue that we should make some sort of drift metric available that you can alert on since detecting manual/out of band changes to production resources is 100% something I would want an alert on

**Comment by sfzylad** [comment]
@dcherman Can you help me to understand why it's out of scope? This scenario is very likely to happen in real world environment and it seems to me e2e testing should consider such scenario too.

**Comment by dcherman** [comment]
@sfzylad I just mean that this particular issue is specific to the testing infrastructure and types of tests that should be implemented.  Discussion about the exact behavior of what should happen when drift occurs absolutely needs to/should happen, but likely in a separate issue.

**Comment by sfzylad** [comment]
@dcherman Cool, thanks for explanation. :)

**Comment by mhausenblas** [comment]
Thanks for all your comments and just to clarify: all above is in scope, it's a design issue, after all. Once the design stabilizes, implementation-level issues will follow, for sure. For now, let's stay open-minded, keep it coming :)

**Comment by sayboras** [comment]
> * Per-AWS-service test: checks to what extent an ASO for a certain service covers the operations supported by said service.

I am just curious about this test, will these tests run with the actual AWS services ? Or is there any plan to use the mock service such as `localstack` for faster feedback? 

Another consideration is about the discrepancies in features that we have in different regions.

**Comment by mhausenblas** [comment]
The tests SHOULD run against the actual APIs and not against local emulations.

**Comment by jaypipes** [comment]
Some more details on a testing plan...

We are going to modify the [e2e integration test setup](https://github.com/aws/amazon-vpc-cni-k8s/blob/bc04604397889430f0a3d5f6e4766b399c1d5fcc/scripts/run-integration-tests.sh) in the AWS VPC CNI plugin repository that uses the [aws-k8s-tester](github.com/aws/aws-k8s-tester) binary for initial setup of the Kubernetes cluster.

Modifications we need to make to this script include:

* Setting up this repository in CircleCI similar to the [configuration in CNI plugin](https://github.com/aws/amazon-vpc-cni-k8s/blob/bc04604397889430f0a3d5f6e4766b399c1d5fcc/.circleci/config.yml)
* Generating an [e2e test](https://github.com/aws/amazon-vpc-cni-k8s/blob/bc04604397889430f0a3d5f6e4766b399c1d5fcc/test/integration/integration_test.go) for simple CRUD+L operations for resources in each generated service controller
* Modify the main [e2e test runner script](https://github.com/aws/amazon-vpc-cni-k8s/blob/bc04604397889430f0a3d5f6e4766b399c1d5fcc/scripts/run-integration-tests.sh) to allow for each service controller binary to be built into its own Docker image and pushed to an ECR repository and installed into the Kubernetes cluster created by `aws-k8s-tester eks create cluster`
* Develop a way for individual service teams to extend the set of integration tests that we generate
* Lots of documentation

**Comment by mhausenblas** [comment]
Can we assign someone?

**Comment by jaypipes** [comment]
Closing this out. We ended up building our own simple Bash-based [integration testing framework](https://github.com/aws/aws-controllers-k8s/blob/5c12617bdbe6c2bec320ac5a0b33ad97e4bd30eb/test/e2e/run-tests.sh) and each service controller has its own set of e2e tests that are run in a manual self-hosted GitHub runner. Unit tests are in Go and included in the individual Go packages.

There is a [separate issue](#440) for a long-running soak test.

And we have [documented release criteria](https://aws.github.io/aws-controllers-k8s/releases/) related to testing of individual service controllers.
